harness/tests/experiment/pytorch_utils.py

```diff
@@ -0,0 +1,265 @@
+# type: ignore
+import os
+import pathlib
+import random
+import sys
+import typing
+import uuid
+
+from torch.distributed import launcher
+
+import determined as det
+from determined import gpu, pytorch
+from tests.experiment import utils
+
+
+def calculate_gradients(
+    batch_size: int = 4, epoch_size: int = 64, num_epochs: int = 3
+) -> typing.List[float]:
+    # independently compute expected metrics
+    batches = [
+        (v[:], v[:])
+        for v in (
+            [x * 0.1 + 1.0 for x in range(y, y + batch_size)]
+            for y in (z % epoch_size for z in range(0, epoch_size * num_epochs, batch_size))
+        )
+    ]
+
+    lr = 0.001
+
+    def compute_expected_weight(
+        data: typing.List[float], label: typing.List[float], w: float
+    ) -> float:
+        n = len(data)
+        expected_step = 2.0 * lr * sum((d * (l - d * w) for d, l in zip(data, label))) / n
+        return w + expected_step
+
+    expected_weights = []
+    weight = 0.0
+    data: typing.List[float] = []
+    label: typing.List[float] = []
+    for i, batch in enumerate(batches):
+        if i % 2 == 0:
+            # for even-numbered batches the optimizer step is a no-op:
+            # the weights don't change
+            data, label = batch
+        else:
+            additional_data, additional_label = batch
+            data += additional_data
+            label += additional_label
+            weight = compute_expected_weight(data, label, weight)
+        expected_weights.append(weight)
+
+    return expected_weights
+
+
+def create_trial_and_trial_controller(

```

Comment:
This is a pretty long method. It could be that I'm just being too lazy, but I don't want to have to read through the whole thing to figure out what it does. If there's anything that's not immediately obvious from its name, could you please drop a docstring into this method?